server/image_pull.go

```diff
@@ -27,38 +27,75 @@ func (s *Server) PullImage(ctx context.Context, req *pb.PullImageRequest) (resp
 		image = img.Image
 	}
 
-	sourceCtx := *s.config.SystemContext // A shallow copy we can modify
+	pullArgs := pullArguments{image: image}
 	if req.GetAuth() != nil {
 		username := req.GetAuth().Username
 		password := req.GetAuth().Password
 		if req.GetAuth().Auth != "" {
 			username, password, err = decodeDockerAuth(req.GetAuth().Auth)
 			if err != nil {
-				log.Debugf(ctx, "error decoding authentication for image %s: %v", img, err)
+				log.Debugf(ctx, "error decoding authentication for image %s: %v", image, err)
 				return nil, err
 			}
 		}
 		// Specifying a username indicates the user intends to send authentication to the registry.
 		if username != "" {
-			sourceCtx.DockerAuthConfig = &types.DockerAuthConfig{
+			pullArgs.credentials = types.DockerAuthConfig{
 				Username: username,
 				Password: password,
 			}
 		}
 	}
 
+	// We use the server's pullCache to record which images are currently being
+	// pulled. This allows for avoiding pulling the same image in parallel.
+	// Hence, if a given image is currently being pulled, we queue into the
+	// pullOperation's waitgroup and wait for the pulling goroutine to unblock
+	// us and re-use its results.
+	pullOp, pullInProcess := s.pullOperation(&pullArgs)
+	if !pullInProcess {
+		go func() {
+			pullOp.imageRef, pullOp.err = s.pullImage(ctx, &pullArgs)
```

Comment:
If you want to deal with `panic`-safety in full:
```go
pullOp.err = errors.New("pullImage was aborted by a Go panic")
defer /* pullOp.wg.Done() etc., under the lock */
pullOp.ImageRef, pullOp.Err = s.pullImage(…)
```
(Or if you decide that it’s not worth it, I’m fine with that as well.)